### PR TITLE
POC #12277 LSP config discussion

### PR DESCRIPTION
**Pull request by landerlo**
- POC that refactors the location queries accepting
a high level callback on_locations to react to the locations
response. This allows for great configurability user config
without involving the setup function.

### CODE REVIEW

**Comment by landerlo** [comment]
A contribution to the discussion in:
https://github.com/neovim/neovim/issues/12277

File: runtime/lua/vim/lsp/buf.lua [hunk 1]
```diff
@@ -153,7 +202,7 @@ end
 --- vim.api.nvim_command [[autocmd CursorMoved <buffer> lua vim.lsp.buf.clear_references()]]
 --- </pre>
 function M.document_highlight()
-  local params = util.make_position_params()
+  local params = util.jkmake_position_params()
```
Comment (clason) [inline runtime/lua/vim/lsp/buf.lua:205]: Is that intended?

Comment (landerlo) [reply runtime/lua/vim/lsp/buf.lua]: Sorry, pushed a wrong commit. It's fixed now

**Comment by tjdevries** [comment]
Thanks @landerlo I will take a look at this soon! (sometime over the next few days. If you don't hear from me, please ping me again on this PR)

**Comment by mfussenegger** [comment]
What do you think of adding just the `on_locations` callback as an additional argument to the `M.references`, `M.definition`, ... functions? And still use the normal `vim.lsp.callbacks` callback when it is not in use. Something like:


```diff
-function M.references(context)
+function M.references(context, on_locations)
   validate { context = { context, 't', true } }
   local params = util.make_position_params()
   params.context = context or {
     includeDeclaration = true;
   }
   params[vim.type_idx] = vim.types.dictionary
-  request('textDocument/references', params)
+  if on_locations then
+    request('textDocument/references', params, mk_locations_callback(on_locations))
+  else
+    request('textDocument/references', params)
+  end
 end
```

I think that would have the advantage that it doesn't break things for users who either already override `vim.lsp.callbacks` or who make use of the current `context` argument.

Otherwise this PR here would also need to extend the documentation in `lsp.txt` to mention some exceptions of `vim.lsp.callbacks` not being used.

Though to be fair, I'm still not sure if it is justified to replace a 

```
if exists('fzf')
  nnoremap <silent> gD <cmd>lua vim.lsp.buf_request(0, 'textDocument/definition', vim.lsp.make_position_params(), metals.on_locations_fuzzy_select)<CR>
endif
```

With

```
if exists('fzf')
  nnoremap <silent> gD <cmd>lua vim.lsp.buf.definition { on_locations=metals.fuzzy_select }<CR>
endif
```

**Comment by landerlo** [comment]
> What do you think of adding just the `on_locations` callback as an additional argument to the `M.references`, `M.definition`, ... functions? And still use the normal `vim.lsp.callbacks` callback when it is not in use. Something like:

Sure, we can keep the old callbacks for backward compatibility, though we are at a time when we can still break things so we should put more weight on what is our preferred approach to simplify things as much as possible. 

Using your api
> +function M.definition(on_locations)

The change would only imply that they would put the function that they were registering in the setup 

```callbacks[textDocument/definition] = custom_fun```
into
```definitions(custom_fun)```
Which I think it's easier to configure and has less interactions and possibilities of errors with other
config and callbacks that do need to be registered in a centralized fashion like edits, publishing or status. But if keeping the compatibility now is a requirement we can keep what we have.

> Otherwise this PR here would also need to extend the documentation in `lsp.txt` to mention some exceptions of `vim.lsp.callbacks` not being used.
You are right this could be a matter of confussion if  so  this is a good argument for keeping the old behaviour as well. No strong feelings either way.
 
> But I'm still not sure if it is justified to replace a
> 
> ```
> if exists('fzf')
>   nnoremap <silent> gD <cmd>lua vim.lsp.buf_request(0, 'textDocument/definition', vim.lsp.make_position_params(), metals.on_locations_fuzzy_select)<CR>
> endif
> ```
> 
> With
> 
> ```
> if exists('fzf')
>   nnoremap <silent> gD <cmd>lua vim.lsp.buf.definition { on_locations=metals.fuzzy_select }<CR>
> endif
> ```
I think the table argument to allow for named parameters makes it nice and self declarative, but this is a personal aesthetic choice, no big deal.

The difference is whether fuzzy_select needs to do the canonicalisation of different formats, table or string and error management, basically copy what's already there in the definitions callback.

If the current M.definitons was broken into a function that does the response and returns a list of locations or passes it to a function it receives (as in my PR) then there would be less copy paste which was part of my goal.

Then it's true that we can call buf_request directly and it's not much longer:
```vim
if exists('fzf')
  nnoremap <silent> gD <cmd>lua vim.lsp.buf_request(0, 'textDocument/definition', vim.lsp.make_position_params(), on_locations(metals.on_locations_fuzzy_select))<CR>
endif
```
Which is still pleasing, but notice the on_locations function that would do the first processing so I can have then
````vim
nnoremap <silent> gp <cmd>lua vim.lsp.buf_request(0, 'textDocument/definition', vim.lsp.make_position_params(), on_locations(vim.lsp.util.preview_location))<CR>
nnoremap <silent> gp <cmd>lua vim.lsp.buf_request(0, 'textDocument/definition', vim.lsp.make_position_params(), on_locations(function(locs) vim.lsp.util.jump_to_location(locs[1])))<CR>
```
And we can have the preview location as it is now without copy pasting any reponse processing, and we can inline functions like there to call a simple reusable function jump_to_location.
It is about promoting the function composition and callback layering technique to avoid the copy paste that could creep configurations with the issue of not benefitting from fixes to the ranges management, etc.

All those low level response processing functions should be decoupled from the final action or UI as much as possible. That would be my main goal being able to reuse those functions and change the UI or action. I'm less bothered by what's the actual signature of vim.lsp.definitions because you are right I can call buf_requests without much more effort.

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/15585 introduced `vim.diagnostic.config`. See for details, but in general the `*.config()` pattern is recommended if generic editor options (`:help options`) are not feasible.

If this PR is still needed please recreate after rebasing.

**Pull request closed without merging**